### DISCUSSION TITLE
Should SWT require Java 17 runtime in 4.28?

### DISCUSSION THREAD

**Discussion by iloveeclipse** [eclipse-platform/eclipse.platform.swt#623] [Ideas]
I've just noticed that SWT is still requires at least Java 11, even if the almost everything in platform (except equinox) is moved to Java 17.

Is there any reason why we didn't switched it yet to Java 17? Note that the [Eclipse project Plan for 4.28](https://www.eclipse.org/projects/project-plan.php?planurl=https://www.eclipse.org/eclipse/development/plans/eclipse_project_plan_4_28.xml) only sees Java 17 as supported platform.

So similar to the discussion on the [ECJ move to Java 17](https://github.com/eclipse-jdt/eclipse.jdt.core/discussions/888), I wonder if we should switch SWT to require Java 17 runtime now.

**The reasons for switch to Java 17 runtime are**:

- **There are no tests running on Java 11 anymore**. So we could have already broken SWT because of some runtime issue we don't see on Java 17 / 19.
- Java 17 is LTS already, and next LTS release (Java 21) is planned for September 2023.
- Use latest Java language features that simplify our life: records, instanceof pattern, etc
- Use latest available JDK libraries, example: [this one line](https://github.com/eclipse-pde/eclipse.pde/blob/8893c7e677e8fb4848e2e75c38e2e259810652e0/ui/org.eclipse.pde.ui.tests/src/org/eclipse/pde/ui/tests/project/ProjectCreationTests.java#L1486) **using Java 17**  was a [complex method with 40 lines](https://github.com/eclipse-pde/eclipse.pde/blob/eaf462c977ca3a1f2e1ef4226f0ba17ec9c8cdab/ui/org.eclipse.pde.ui.tests/src/org/eclipse/pde/ui/tests/project/ProjectCreationTests.java#L1486) in Java 8!
- Simplify our life maintaining SWT source code configuration, build, test etc - all the jenkins/maven/github/ whatever configurations where we should simply **reuse** solutions we would have with default platform improvements we see every day.
- basically: **save our time**. 
- Our development work would only benefit from Java 17.
- Switching to Java 17 now could shorten the time for possible later transition to next Java that would include [project Panama API](https://openjdk.org/jeps/424).
- Consumers that need SWT running on Java 11 can use older SWT versions.

**The reasons for stay on Java 11 runtime are**:
- Java 11 is still an active version that is still being supported by Oracle. Platform moving ahead is a complex case, but there's no such reason to do that in the SWT library case.
- We should take a moment to think about all the SWT consumers we will either be leaving behind or upsetting.
- Java 17 is fairly new, by Java standards and only a year and half old. Those clients that prefer to go with tried and tested software will not be happy about this.

**Comment by mickaelistria** [comment]
+1 to move SWT to Java 17.

**Comment by akurtakov** [comment]
I don't mind moving to Java 17 as I see Java 11 as self support even now. Even  Java 17 is gonna be self support as soon as Java 21 is out in few months from my side.

**Comment by vogella** [comment]
+1 to move SWT to Java 17.

**Comment by iloveeclipse** [comment]
@SyntevoAlex : I know Syntevo uses SWT in standalone mode, any concerns from your side?

**Comment by SyntevoAlex** [comment]
We already use Java 17.
Thanks for asking!

**Comment by vogella** [comment]
I suggest to go ahead with this change. @iloveeclipse not sure if we need another email to cross, we already announced the move to Java 17 so I think you can update SWT. Would be good to have this done before M2 so that others can see this change and if requried can react to it.

**Reply by iloveeclipse** [reply]
I've sent a mail to cross project list and will propose a PR.

**Comment by iloveeclipse** [comment]
See https://github.com/eclipse-platform/eclipse.platform.swt/pull/627. I plan to merge today & see what will break in SDK build.

**Comment by iloveeclipse** [comment]
Ok, with the build https://download.eclipse.org/eclipse/downloads/drops4/I20230413-0320/ SWT requires Java 17.
